crypto/zk/prover/prover.go

```diff
@@ -21,10 +21,59 @@ var (
 	ErrWitnessCalc     = fmt.Errorf("error during witness calculation")
 	ErrProofGen        = fmt.Errorf("error during zksnark proof generation")
 	ErrParseProofData  = fmt.Errorf("error parsing the proof provided, it must be a valid json, check https://github.com/iden3/go-rapidsnark/blob/73d5784d2aa791dd6646142b0017dbef97240f57/types/proof.go")
-	ErrParsePubSignals = fmt.Errorf("error during zksnark proof generation, it must be a json with array of strings")
+	ErrParsePubSignals = fmt.Errorf("error during zksnark public signals proof generation, it must be a json with array of strings")
+	ErrEncodingProof   = fmt.Errorf("error encoding prove result into a proof struct")
+	ErrDecodingProof   = fmt.Errorf("error decoding prove as []byte")
 	ErrVerifyProof     = fmt.Errorf("error during zksnark verification")
 )
 
+// ProofData struct contains the calculated parameters of a Proof. It allows to
+// encode and decode go-rapidsnark inputs and outputs easily.
+type ProofData struct {
+	A []string   `json:"pi_a"`
+	B [][]string `json:"pi_b"`
+	C []string   `json:"pi_c"`
+}
+
+// Proof struct wraps the ProofData struct and its associated public signals.
+// Contains all the required information to perform a proof verification.
+type Proof struct {
+	Data       ProofData `json:"data"`
+	PubSignals []string  `json:"pubSignals"`
+}
+
+// ParseProof function encodes the provided proof data and public signals into a
+// Proof struct, performing an unmarshal operation over them. Returns an error
+// if something is wrong.
+func ParseProof(proofData, pubSignals []byte) (*Proof, error) {
+	data := ProofData{}
+	if err := json.Unmarshal(proofData, &data); err != nil {
+		return nil, ErrEncodingProof
+	}
+
+	signals := []string{}
+	if err := json.Unmarshal(pubSignals, &signals); err != nil {
+		return nil, ErrEncodingProof
+	}
+	return &Proof{Data: data, PubSignals: signals}, nil
+}
+
+// Bytes function returns the current Proof struct parameters Data and
+// PubSignals as []byte. It returns an error if something fails.
+func (p *Proof) Bytes() (proofData []byte, pubSignals []byte, err error) {
```

Comment:
We usually prefer not to use named return parameters. It can be sometimes confusing for someone reading or modifying the code. Declare variables in the method body is much more direct and might avoid some mistakes.